torch/csrc/autograd/python_variable.cpp

```diff
@@ -1122,6 +1135,124 @@ static PyObject* DTensor_OpSchema_post_init(PyObject* mod, PyObject* self) {
   END_HANDLE_TH_ERRORS
 }
 
+static py::list int_array_to_list(IntArrayRef arr) {
+  py::list result(arr.size());
+  for (const auto idx : c10::irange(arr.size())) {
+    result[idx] = arr[idx];
+  }
+  return result;
+}
+
+static PyObject* DTensor_compute_global_tensor_info_impl(
+    const Tensor& tensor,
+    py::handle mesh,
+    const py::sequence& placements) {
+  Py_ssize_t idx = 0;
+  DimVector tensor_shape(tensor.sizes().begin(), tensor.sizes().end());
+  DimVector tensor_strides(tensor.strides().begin(), tensor.strides().end());
+  // NOTE: if this is a py::handle then this code stops working;
+  // apparently we can't rely on the bound method to stick around.
+  py::object mesh_size;
+  for (const auto& placement : placements) {
+    // TODO: C++ify Placement and DeviceMesh somehow; profiling seems
+    // to say that nearly all our remaining time spent is spent
+    // calling back into Python.
+    if (checked_is_true(
+            placement.attr(dtensor_interned_strings.is_shard)().ptr())) {
+      const auto shard_dim =
+          py::cast<int64_t>(placement.attr(dtensor_interned_strings.dim));
+      TORCH_CHECK(
```

Comment:
Don't we have a TORCH_CHECK_ASSERT to raise an Assertion Error?

Reply:
not that I found. TORCH_INTERNAL_ASSERT exists but doesn't seem to do anything differently w.r.t. the exception it raises.